File: mParticle-Apple-SDK/Utils/MPMessageBuilder.m
#import "MPMessageBuilder.h"
#import "MPStateMachine.h"
#import <UIKit/UIKit.h>
#import "MPEnums.h"
#import "MPCommerceEvent.h"
#import "MPCommerceEvent+Dictionary.h"
#import "MPILogger.h"
#import "NSDictionary+MPCaseInsensitive.h"
#import "MPPersistenceController.h"
#import "mParticle.h"

@import mParticle_Apple_SDK_Swift;

NSString *const launchInfoStringFormat = @"%@%@%@=%@";
NSString *const kMPHorizontalAccuracyKey = @"acc";
NSString *const kMPLatitudeKey = @"lat";
NSString *const kMPLongitudeKey = @"lng";
NSString *const kMPVerticalAccuracyKey = @"vacc";
NSString *const kMPRequestedAccuracy = @"racc";
NSString *const kMPDistanceFilter = @"mdst";
NSString *const kMPIsForegroung = @"fg";
NSString *const kMPUserAttributeWasDeletedKey = @"d";
NSString *const kMPUserAttributeNewValueKey = @"nv";
NSString *const kMPUserAttributeOldValueKey = @"ov";
NSString *const kMPUserAttributeNewlyAddedKey = @"na";
NSString *const kMPUserIdentityNewValueKey = @"ni";
NSString *const kMPUserIdentityOldValueKey = @"oi";

@interface MParticle ()
@property (nonatomic, strong, readonly) MPStateMachine_PRIVATE *stateMachine;
@property (nonatomic, strong, nullable) NSString *dataPlanId;
@property (nonatomic, strong, nullable) NSNumber *dataPlanVersion;
@end

@interface MPMessageBuilder ()
@property (atomic, strong) NSMutableDictionary<NSString *, id> *messageDictionary;
@property (nonatomic, copy) NSString *uuid;
@property (nonatomic) MPMessageType messageTypeValue;
@end

@implementation MPMessageBuilder

+ (NSString *)stringForMessageType:(MPMessageType)type {
    NSString *string = [MPMessageBuilderFields stringForMessageTypeRawValue:type];
    if (!string) {
        MPILogError(@"Unknown message type enum: %@", @(type));
        return kMPMessageTypeStringUnknown;
    }

    return string;
}

+ (MPMessageType)messageTypeForString:(NSString *)string {
    NSNumber *rawType = [MPMessageBuilderFields rawMessageTypeForString:string];
    if (rawType == nil) {
        MPILogError(@"Unknown message type string: %@", string);
        return MPMessageTypeUnknown;
    }

    return (MPMessageType)rawType.unsignedIntegerValue;
}

- (instancetype)initWithMessageType:(MPMessageType)messageType session:(MPSession *)session {
    self = [super init];
    if (!self || !messageType) {
        return nil;
    }
    
    _timestamp = [[NSDate date] timeIntervalSince1970];
    _messageDictionary = [[NSMutableDictionary alloc] initWithCapacity:5];
    _messageDictionary[kMPTimestampKey] = MPMilliseconds(_timestamp);
    
    _messageTypeValue = messageType;
    _messageType = [MPMessageBuilder stringForMessageType:messageType];
    
    _session = session;
    if (session) {
        if (messageType == MPMessageTypeSessionStart) {
            _uuid = _session.uuid;
        } else {
            _messageDictionary[kMPSessionIdKey] = _session.uuid;
            _messageDictionary[kMPSessionStartTimestamp] = MPMilliseconds(_session.startTime);
            
            if (messageType == MPMessageTypeSessionEnd) {
                NSArray<NSNumber *> *userIdNumbers = [MPMessageBuilderFields filteredUserIdsFrom:_session.sessionUserIds];
                if (userIdNumbers) {
                    _messageDictionary[kMPSessionUserIdsKey] = userIdNumbers;
                }
            }
        }
    }
    
    _dataPlanId = [MParticle sharedInstance].dataPlanId;
    _dataPlanVersion = [MParticle sharedInstance].dataPlanVersion;
    
    NSString *presentedViewControllerDescription = nil;
    NSNumber *mainThreadFlag;
    if ([NSThread isMainThread]) {
        if (![MPStateMachine_PRIVATE isAppExtension]) {
            UIWindow *keyWindow = nil;
            
            // Get key window from active window scene (iOS 13+)
            NSSet<UIScene *> *connectedScenes = [MPApplication_PRIVATE sharedUIApplication].connectedScenes;
            for (UIScene *scene in connectedScenes) {
                if ([scene isKindOfClass:[UIWindowScene class]] && scene.activationState == UISceneActivationStateForegroundActive) {
                    UIWindowScene *windowScene = (UIWindowScene *)scene;
                    for (UIWindow *window in windowScene.windows) {
                        if (window.isKeyWindow) {
                            keyWindow = window;
                            break;
                        }
                    }
                    if (keyWindow) {
                        break;
                    }
                }
            }
            
            UIViewController *presentedViewController = keyWindow.rootViewController.presentedViewController;
            presentedViewControllerDescription = presentedViewController ? [[presentedViewController class] description] : nil;
        } else {
            presentedViewControllerDescription = @"extension_message";
        }
        
        mainThreadFlag = @YES;
    } else {
        presentedViewControllerDescription = @"off_thread";
        mainThreadFlag = @NO;
    }
    
    if (presentedViewControllerDescription) {
        _messageDictionary[kMPPresentedViewControllerKey] = presentedViewControllerDescription;
    }
    _messageDictionary[kMPMainThreadKey] = mainThreadFlag;
    
    return self;
}

- (instancetype)initWithMessageType:(MPMessageType)messageType session:(MPSession *)session messageInfo:(NSDictionary<NSString *, id> *)messageInfo {
    self = [self initWithMessageType:messageType session:session];
    if (self && messageInfo) {
        [_messageDictionary addEntriesFromDictionary:messageInfo];
        
        NSDictionary *messageAttributes = _messageDictionary[kMPAttributesKey];
        if (messageAttributes) {
            _messageDictionary[kMPAttributesKey] = [messageAttributes transformValuesToString];
        }
    }
    return self;
}

- (instancetype)initWithMessageType:(MPMessageType)messageType session:(MPSession *)session userIdentityChange:(MPUserIdentityChangePRIVATE *)userIdentityChange {
    self = [self initWithMessageType:messageType session:session];
    if (self && userIdentityChange) {
        [self userIdentityChange:userIdentityChange];
    }
    return self;
}

- (instancetype)initWithMessageType:(MPMessageType)messageType session:(MPSession *)session userAttributeChange:(MPUserAttributeChange *)userAttributeChange {
    self = [self initWithMessageType:messageType session:session];
    if (self && userAttributeChange) {
        [self userAttributeChange:userAttributeChange];
    }
    return self;
}

#pragma mark Private methods
- (void)userAttributeChange:(nonnull MPUserAttributeChange *)userAttributeChange {
    id oldValue = userAttributeChange.userAttributes[userAttributeChange.key];
    MPUserAttributeChangeFields *fields = [MPMessageBuilderFields userAttributeChangeFieldsWithDeleted:userAttributeChange.deleted
                                                                                                     key:userAttributeChange.key
                                                                                                oldValue:oldValue
                                                                                                newValue:userAttributeChange.valueToLog];

    _messageDictionary[kMPUserAttributeWasDeletedKey] = @(fields.deleted);
    _messageDictionary[kMPEventNameKey] = fields.attributeKey;
    _messageDictionary[kMPUserAttributeOldValueKey] = fields.oldValue;
    _messageDictionary[kMPUserAttributeNewValueKey] = fields.newValue;
    _messageDictionary[kMPUserAttributeNewlyAddedKey] = @(fields.newlyAdded);
}

- (void)userIdentityChange:(MPUserIdentityChangePRIVATE *)userIdentityChange {
    NSDictionary *dictionary = [userIdentityChange.newUserIdentity dictionaryRepresentation];
    if (dictionary) {
        _messageDictionary[kMPUserIdentityNewValueKey] = dictionary;
    }
    
    dictionary = [userIdentityChange.oldUserIdentity dictionaryRepresentation];
    if (dictionary) {
        _messageDictionary[kMPUserIdentityOldValueKey] = dictionary;
    }
}

#pragma mark Public instance methods
- (NSDictionary *)messageInfo {
    return _messageDictionary;
}

- (void)launchInfo:(NSDictionary *)launchInfo {
    NSString *launchInfoString = [MPMessageBuilderFields launchInfoStringFrom:launchInfo];
    if (launchInfoString) {
        _messageDictionary[kMPLaunchURLKey] = launchInfoString;
    }
}

- (void)timestamp:(NSTimeInterval)timestamp {
    _timestamp = timestamp;
    _messageDictionary[kMPTimestampKey] = MPMilliseconds(_timestamp);
}

// NOTE: Here "sessionFinalized" is really referring to if we are starting a new session on launch, see Facebook event forwarder backend code
- (void)stateTransition:(BOOL)sessionFinalized previousSession:(MPSession *)previousSession {
    MPStateMachine_PRIVATE *stateMachine = [MParticle sharedInstance].stateMachine;

    MPStateTransitionFields *fields = [MPMessageBuilderFields stateTransitionFieldsWithSourceApplication:stateMachine.launchInfo.sourceApplication
                                                                                           launchURLString:stateMachine.launchInfo.url.absoluteString
                                                                                          launchParameters:stateMachine.launchInfo.annotation
                                                                             previousSessionInterruptions:previousSession ? previousSession.numberOfInterruptions : 0
                                                                                         sessionFinalized:sessionFinalized];

    if (fields.sourceApplication) {
        _messageDictionary[kMPLaunchSourceKey] = fields.sourceApplication;
    }

    if (fields.launchURLString) {
        _messageDictionary[kMPLaunchURLKey] = fields.launchURLString;
    }

    if (fields.launchParameters) {
        _messageDictionary[kMPLaunchParametersKey] = fields.launchParameters;
    }

    _messageDictionary[kMPLaunchNumberOfSessionInterruptionsKey] = @(fields.numberOfSessionInterruptions);
    _messageDictionary[kMPLaunchSessionFinalizedKey] = @(fields.sessionFinalized);
}

- (MPMessage *)build {
    _messageDictionary[kMPMessageTypeKey] = _messageType;
    _messageDictionary[kMPMessageIdKey] = _uuid ?: [[NSUUID UUID] UUIDString];
    
    NSNumber *userId = _session.userId.integerValue ? _session.userId : [MPPersistenceController_PRIVATE mpId];

    MPMessage *message = [[MPMessage alloc] initWithSession:_session
                                                messageType:_messageType
                                                messageInfo:[_messageDictionary copy]
                                               uploadStatus:MPUploadStatusBatch
                                                       UUID:_messageDictionary[kMPMessageIdKey]
                                                  timestamp:_timestamp
                                                     userId:userId
                                                 dataPlanId:_dataPlanId
                                            dataPlanVersion:_dataPlanVersion];
    return message;
}

@end

File: mParticle-Apple-SDK-Swift/Sources/Utils/MPMessageBuilder.swift
import Foundation
#if canImport(UIKit)
    import UIKit
#endif

/// The values `MPMessageBuilderPRIVATE` needs from `MParticle.sharedInstance`, which the Swift
/// module cannot import. The Objective-C boundary fills it and passes it in, mirroring the
/// `MPURLRequestContext` value-context pattern.
@objc(MPMessageBuilderContext)
public final class MPMessageBuilderContext: NSObject {
    @objc public let dataPlanId: String?
    @objc public let dataPlanVersion: NSNumber?
    @objc public let logger: MPLog?

    @objc public init(dataPlanId: String?, dataPlanVersion: NSNumber?, logger: MPLog?) {
        self.dataPlanId = dataPlanId
        self.dataPlanVersion = dataPlanVersion
        self.logger = logger
        super.init()
    }
}

/// Assembles the message dictionary an `MPMessage` is built from.
///
/// The Objective-C `MPMessageBuilder` this replaces held an `atomic` pointer to a
/// `NSMutableDictionary` whose *contents* were mutated without synchronization, so the atomicity
/// never protected the dictionary itself. A builder is created, mutated and built by one owner on
/// `[MParticle messageQueue]`, so a plain Swift dictionary is equivalent.
@objc(MPMessageBuilder)
public final class MPMessageBuilderPRIVATE: NSObject {
    private var messageDictionary: [AnyHashable: Any]
    private var uuid: String?
    private let context: MPMessageBuilderContext

    @objc public private(set) var messageType: String
    @objc public private(set) var session: MPSessionPRIVATE?
    @objc public private(set) var timestamp: TimeInterval
    @objc public private(set) var dataPlanId: String?
    @objc public private(set) var dataPlanVersion: NSNumber?

    @objc public var messageInfo: [AnyHashable: Any] { messageDictionary }

    // MARK: - Message type <-> wire string

    /// `nil` from `MPMessageBuilderFields` means the raw value is not a known message type; the
    /// error is logged and `kMPMessageTypeStringUnknown` returned, as the original `default:`
    /// branch did.
    @objc(stringForMessageType:logger:)
    public static func string(forMessageType rawValue: UInt, logger: MPLog?) -> String {
        guard let string = MPMessageBuilderFields.string(forMessageTypeRawValue: rawValue) else {
            logger?.error("Unknown message type enum: \(rawValue)")
            return kMPMessageTypeStringUnknown
        }
        return string
    }

    /// Returns the raw `MPMessageType` value, or `MPMessageTypeUnknown` (0) when the string
    /// matches none of the known constants — again logging first, as the original did.
    @objc(messageTypeForString:logger:)
    public static func messageType(forString string: String, logger: MPLog?) -> UInt {
        guard let rawType = MPMessageBuilderFields.rawMessageType(forString: string) else {
            logger?.error("Unknown message type string: \(string)")
            return 0
        }
        return rawType.uintValue
    }

    // MARK: - Init

    @objc(initWithMessageType:session:context:)
    public init?(messageType rawValue: UInt, session: MPSessionPRIVATE?, context: MPMessageBuilderContext) {
        // The original returned nil for `MPMessageTypeUnknown`, whose raw value is 0.
        guard rawValue != 0 else { return nil }

        self.context = context
        self.session = session
        timestamp = Date().timeIntervalSince1970
        messageType = Self.string(forMessageType: rawValue, logger: context.logger)
        dataPlanId = context.dataPlanId
        dataPlanVersion = context.dataPlanVersion

        messageDictionary = [MessageKeys.kMPTimestampKey: MPMilliseconds(timestamp: timestamp)]

        super.init()

        if let session {
            if rawValue == MPMessageTypeSwift.sessionStart.rawValue {
                uuid = session.uuid
            } else {
                messageDictionary[MessageKeys.kMPSessionIdKey] = session.uuid
                messageDictionary[MessageKeys.kMPSessionStartTimestamp] = MPMilliseconds(timestamp: session.startTime)

                if rawValue == MPMessageTypeSwift.sessionEnd.rawValue {
                    let userIdNumbers = MPMessageBuilderFields.filteredUserIds(from: session.sessionUserIds)
                    messageDictionary[kMPSessionUserIdsKey] = userIdNumbers
                }
            }
        }

        applyPresentationContext()
    }

    @objc(initWithMessageType:session:messageInfo:context:)
    public convenience init?(
        messageType rawValue: UInt,
        session: MPSessionPRIVATE?,
        messageInfo: [AnyHashable: Any]?,
        context: MPMessageBuilderContext
    ) {
        self.init(messageType: rawValue, session: session, context: context)
        guard let messageInfo else { return }

        messageDictionary.merge(messageInfo) { _, new in new }

        // The original messaged `transformValuesToString` on whatever sat under the attributes
        // key; a non-dictionary would have crashed. Leaving it untouched is the same for every
        // valid payload and safer for an invalid one.
        if let messageAttributes = messageDictionary[MessageKeys.kMPAttributesKey] as? NSDictionary {
            messageDictionary[MessageKeys.kMPAttributesKey] = AttributeValueTransformer
                .transformedAttributeValues(in: messageAttributes, logger: context.logger)
        }
    }

    @objc(initWithMessageType:session:userIdentityChange:context:)
    public convenience init?(
        messageType rawValue: UInt,
        session: MPSessionPRIVATE?,
        userIdentityChange: MPUserIdentityChangePRIVATE?,
        context: MPMessageBuilderContext
    ) {
        self.init(messageType: rawValue, session: session, context: context)
        guard let userIdentityChange else { return }
        apply(userIdentityChange: userIdentityChange)
    }

    @objc(initWithMessageType:session:userAttributeChange:context:)
    public convenience init?(
        messageType rawValue: UInt,
        session: MPSessionPRIVATE?,
        userAttributeChange: MPUserAttributeChange?,
        context: MPMessageBuilderContext
    ) {
        self.init(messageType: rawValue, session: session, context: context)
        guard let userAttributeChange else { return }
        apply(userAttributeChange: userAttributeChange)
    }

    // MARK: - Mutators

    @objc(timestamp:)
    public func updateTimestamp(_ timestamp: TimeInterval) {
        self.timestamp = timestamp
        messageDictionary[MessageKeys.kMPTimestampKey] = MPMilliseconds(timestamp: timestamp)
    }

    /// NOTE: here "sessionFinalized" is really referring to whether we are starting a new session
    /// on launch, see Facebook event forwarder backend code.
    @objc(stateTransition:previousSession:launchInfo:)
    public func stateTransition(
        _ sessionFinalized: Bool,
        previousSession: MPSessionPRIVATE?,
        launchInfo: MPLaunchInfo?
    ) {
        let fields = MPMessageBuilderFields.stateTransitionFields(
            sourceApplication: launchInfo?.sourceApplication,
            launchURLString: launchInfo?.url.absoluteString,
            launchParameters: launchInfo?.annotation,
            previousSessionInterruptions: previousSession.map { Int($0.numberOfInterruptions) } ?? 0,
            sessionFinalized: sessionFinalized
        )

        if let sourceApplication = fields.sourceApplication {
            messageDictionary[kMPLaunchSourceKey] = sourceApplication
        }
        if let launchURLString = fields.launchURLString {
            messageDictionary[kMPLaunchURLKey] = launchURLString
        }
        if let launchParameters = fields.launchParameters {
            messageDictionary[kMPLaunchParametersKey] = launchParameters
        }
        messageDictionary[kMPLaunchNumberOfSessionInterruptionsKey] = NSNumber(value: fields.numberOfSessionInterruptions)
        messageDictionary[kMPLaunchSessionFinalizedKey] = NSNumber(value: fields.sessionFinalized)
    }

    // MARK: - Build

    @objc public func build() -> MPMessagePRIVATE {
        messageDictionary[MessageKeys.kMPMessageTypeKey] = messageType
        let messageId = uuid ?? UUID().uuidString
        messageDictionary[MessageKeys.kMPMessageIdKey] = messageId

        let sessionUserId = session?.userId
        let userId = (sessionUserId?.intValue ?? 0) != 0 ? sessionUserId! : MPUserDefaults.storedMpId()

        return MPMessagePRIVATE(
            session: session,
            messageType: messageType,
            messageInfo: messageDictionary as NSDictionary,
            uploadStatus: kMPUploadStatusBatch,
            uuid: messageId,
            timestamp: timestamp,
            userId: userId,
            dataPlanId: dataPlanId,
            dataPlanVersion: dataPlanVersion
        )
    }

    // MARK: - Private

    private func apply(userAttributeChange: MPUserAttributeChange) {
        let oldValue = userAttributeChange.userAttributes?[userAttributeChange.key]
        let fields = MPMessageBuilderFields.userAttributeChangeFields(
            deleted: userAttributeChange.deleted,
            key: userAttributeChange.key,
            oldValue: oldValue,
            newValue: userAttributeChange.valueToLog
        )

        messageDictionary[kMPUserAttributeWasDeletedKey] = NSNumber(value: fields.deleted)
        messageDictionary[MessageKeys.kMPEventNameKey] = fields.attributeKey
        messageDictionary[kMPUserAttributeOldValueKey] = fields.oldValue
        messageDictionary[kMPUserAttributeNewValueKey] = fields.newValue
        messageDictionary[kMPUserAttributeNewlyAddedKey] = NSNumber(value: fields.newlyAdded)
    }

    private func apply(userIdentityChange: MPUserIdentityChangePRIVATE) {
        if let dictionary = userIdentityChange.newUserIdentity?.dictionaryRepresentation() {
            messageDictionary[kMPUserIdentityNewValueKey] = dictionary
        }
        if let dictionary = userIdentityChange.oldUserIdentity?.dictionaryRepresentation() {
            messageDictionary[kMPUserIdentityOldValueKey] = dictionary
        }
    }

    /// Records which view controller was on screen and whether the message was built on the main
    /// thread, matching the original's three-way outcome.
    private func applyPresentationContext() {
        let presentedViewControllerDescription: String?
        let isMainThread = Thread.isMainThread

        if isMainThread {
            if MPStateMachinePRIVATE.isAppExtension() {
                presentedViewControllerDescription = "extension_message"
            } else {
                presentedViewControllerDescription = Self.presentedViewControllerDescription()
            }
        } else {
            presentedViewControllerDescription = "off_thread"
        }

        if let presentedViewControllerDescription {
            messageDictionary[kMPPresentedViewControllerKey] = presentedViewControllerDescription
        }
        messageDictionary[kMPMainThreadKey] = NSNumber(value: isMainThread)
    }

    private static func presentedViewControllerDescription() -> String? {
        guard let application = MPApplication_PRIVATE.sharedUIApplication() else { return nil }

        // `connectedScenes` is imported as non-optional, but the Objective-C getter can still
        // answer nil — an OCMock of `UIApplication` with the selector unstubbed does, and the
        // previous Objective-C code simply iterated an empty set. Read it through the runtime so
        // a nil answer stays nil instead of trapping.
        let connectedScenes = application
            .perform(NSSelectorFromString("connectedScenes"))?
            .takeUnretainedValue() as? Set<UIScene> ?? []

        // Key window from the active window scene (iOS 13+).
        let keyWindow = connectedScenes
            .lazy
            .compactMap { $0 as? UIWindowScene }
            .filter { $0.activationState == .foregroundActive }
            .compactMap { scene in scene.windows.first(where: \.isKeyWindow) }
            .first

        guard let presented = keyWindow?.rootViewController?.presentedViewController else {
            return nil
        }
        // The original used `[[presentedViewController class] description]`, i.e. the Objective-C
        // runtime name, which is not what Swift's `type(of:)` prints for a Swift class.
        return NSStringFromClass(type(of: presented))
    }
}

// Wire keys mirrored from `MPIConstants.m` and the deleted `MPMessageBuilder.{h,m}`; the Swift
// module cannot import the ObjC module, so the canonical values are duplicated here as literals.
private let kMPMessageTypeStringUnknown = "unknown"
private let kMPSessionUserIdsKey = "smpids"
private let kMPPresentedViewControllerKey = "vc"
private let kMPMainThreadKey = "mt"
private let kMPLaunchSourceKey = "src"
private let kMPLaunchURLKey = "lr"
private let kMPLaunchParametersKey = "lpr"
private let kMPLaunchSessionFinalizedKey = "sf"
private let kMPLaunchNumberOfSessionInterruptionsKey = "nsi"
private let kMPUserAttributeWasDeletedKey = "d"
private let kMPUserAttributeNewValueKey = "nv"
private let kMPUserAttributeOldValueKey = "ov"
private let kMPUserAttributeNewlyAddedKey = "na"
private let kMPUserIdentityNewValueKey = "ni"
private let kMPUserIdentityOldValueKey = "oi"

/// `MPUploadStatusBatch` (`MPIConstants.h`).
private let kMPUploadStatusBatch = 1
